modeling.py

```diff
@@ -0,0 +1,146 @@
+        '''
+        Output: DataFrame - index=player_ids, single column of data corresponding to the latent
+                feature from the nmf model relating to player skill
+        '''
+        position_skill = pd.DataFrame(self.model.basis().toarray(), 
+                                      index=self.player_ids, 
+                                      columns=['skill'])
+        return position_skill
+
+    def get_defense_skill(self):
+        '''
+        Output: DataFrame - index=opponent (aka defense), single column of data corresponding 
+                to the latent feature from the nmf model relating the defense's inability to 
+                defend against the position
+        '''
+        defense_skill = pd.DataFrame(self.model.coef().toarray().T,
+                                     index=self.opponents, 
+                                     columns=['swiss']) 
+
+        return defense_skill
+
+    def view_sorted_position_skill(self):
+        '''
+        Output: DataFrame - only has names of players and 'skill' column from NMF to eyeball
+                efficacy of model
+        '''
+        # Prepare to merge back with player names for human readability
+        names_columns = ['player_id', 'full_name']
+        position_names = self.df[names_columns]
+
+        skill_names =  position_names.merge(right=self.get_position_skill(),
+                                            how='inner',
+                                            right_index=True,
+                                            left_on='player_id') 
+
+        skills = skill_names.groupby(names_columns).first()
+
+        return skills.sort(columns=['skill'], ascending=False)
+
+    def view_sorted_defense_skill(self):
+        '''
+        Output: DataFrame - only has defense name and 'swiss' columns, names for how like swiss
+                cheese the defense is against the position
+        '''
+        return self.get_defense_skill().sort('swiss')
+
+class PositionNMFFactory(object):
+    '''
+    Class for decomposing nfl data with non-negative factorization
+    ''' 
+    def __init__(self, year, till_week):
+        self.year = year
+        self.till_week = till_week
+        self.nfl_frames = NFLFrames()
+        self.make_year_till_week()
+
+    def make_year_till_week(self):
+        '''
+        Output: DataFrame - players of given position, their player_id, and the corresponding 
+                fanduel points and opponent those points were scored against
+        '''
+        def get_year_week(week):
+            '''
+            Helper function to grab the player_id, opponent, and fanduel points for a position 
+            in the given year until max_week
+            '''
+            return self.nfl_frames.get_year_week_frame(self.year, week)
+
+        weeks = range(1, self.till_week + 1)
+        self.year_till_week = pd.concat([get_year_week(week) for week in weeks], axis=0)
+
+    def get_position_model(self, position): 
```

Comment:
On further reflection, it doesn't look like you are returning the defense strengths against each position. Not sure if I had your reason right for why we are doing this by position, but I think you want to keep track of defense strengths